### PR TITLE
DPR2-123: Upsert incremental domain inserts

### PR DESCRIPTION
**Pull request by koladeadewuyi-moj**

### CODE REVIEW

File: src/main/java/uk/gov/justice/digital/domain/DomainExecutor.java [hunk 1]
```diff
@@ -266,7 +269,7 @@ public void applyDomain(
                 storage.endTableUpdates(spark, target);
                 break;
             case Insert:
-                storage.append(target.toPath(), dataFrame);
+                storage.upsertRecords(spark, target.toPath(), dataFrame, primaryKey);
```
Comment (koladeadewuyi-moj) [inline src/main/java/uk/gov/justice/digital/domain/DomainExecutor.java:272]: This fixes the issue of duplicate records by performing an upsert of the record

File: src/main/java/uk/gov/justice/digital/service/DataStorageService.java [hunk 2]
```diff
@@ -81,47 +81,80 @@ public void appendDistinct(String tablePath, Dataset<Row> df, SourceReference.Pr
...
+                        .merge(dataFrame.as(TARGET), condition)
+                        .whenMatched()
+                        .updateAll()
+                        .execute();
+            } else {
+                logger.error("Failed to update table {}. Delta table is not present", tablePath);
+            }
+        } catch (Exception e) {
+            val errorMessage = String.format("Failed to update table %s", tablePath);
+            logger.error(errorMessage, e);
```
Comment (koladeadewuyi-moj) [inline src/main/java/uk/gov/justice/digital/service/DataStorageService.java:132]: Only logging storage errors for CDC events so the streaming job proceeds.

File: src/main/java/uk/gov/justice/digital/writer/Writer.java
Comment (dominic-messenger-moj) [inline src/main/java/uk/gov/justice/digital/writer/Writer.java:1]: This has been a longstanding bugbear of mine. Glad it has been resolved.

**Review by dominic-messenger-moj — APPROVED** [review]
LGTM

**Pull request merged**